src/fcc_bridge_node_ros.cpp

```diff
@@ -0,0 +1,225 @@
+    rclcpp::SubscriptionOptions subscription_options;
+    subscription_options.content_filter_options.filter_expression =
+        "sender_id = 'mission_control'";
+
+    // Create Heartbeat subscription
+    this->mission_control_heartbeat_subscriber =
+        this->create_subscription<interfaces::msg::Heartbeat>(
+            "heartbeat", 1,
+            std::bind(&FCCBridgeNode::mission_control_heartbeat_subscriber_cb,
+                      this, std::placeholders::_1),
+            subscription_options);
+
+    // Ensure that the filter is enabled. (Some DDS versions do not support it)
+    if (!this->mission_control_heartbeat_subscriber->is_cft_enabled()) {
+        RCLCPP_FATAL(this->get_logger(), "Content filtering is not enabled!");
+        this->internal_state = INTERNAL_STATE::ERROR;
+        return;
+    }
+
+    // Setup 5Hz timer to get telemetry from the FCC
+    this->fcc_telemetry_timer_5hz = this->create_wall_timer(
+        std::chrono::milliseconds{200},
+        std::bind(&FCCBridgeNode::fcc_telemetry_timer_5hz_cb, this));
+
+    // Setup 10Hz timer to get telemetry from the FCC
+    this->fcc_telemetry_timer_5hz = this->create_wall_timer(
+        std::chrono::milliseconds{100},
+        std::bind(&FCCBridgeNode::fcc_telemetry_timer_10hz_cb, this));
+}
+
+void FCCBridgeNode::mission_control_heartbeat_subscriber_cb(
+    const interfaces::msg::Heartbeat &msg) {
+    RCLCPP_DEBUG(this->get_logger(), "Received heartbeat from mission control");
+
+    // Check if the heartbeat tick has increased since last time
+    if (msg.tick <= this->last_mission_control_heartbeat.tick) {
+        RCLCPP_ERROR(this->get_logger(),
+                     "The received heartbeat is not newer than the last one! "
+                     "Triggering RTH...");
+        this->trigger_rth();
+    }
+
+    // Check if mission control is still active
+    if (!msg.active) {
+        RCLCPP_ERROR(this->get_logger(),
+                     "Mission control node is inactive! Triggering RTH...");
+        this->trigger_rth();
+    }
+
+    // Set the cached heartbeat message to the current one
+    this->last_mission_control_heartbeat = msg;
+
+    // Verify that the received heartbeat is not too old
+    this->check_last_mission_control_heatbeat();
+
+    RCLCPP_INFO(this->get_logger(), "Mission control is alive and ok");
+}
+
+void FCCBridgeNode::fcc_telemetry_timer_5hz_cb() {
+    RCLCPP_DEBUG(this->get_logger(),
+                 "5Hz telemetry timer callback was triggered");
+    this->check_last_mission_control_heatbeat();
+    switch (this->internal_state) {
+        case INTERNAL_STATE::STARTING_UP:
+        case INTERNAL_STATE::ROS_SET_UP:
+        case INTERNAL_STATE::ERROR:
+            RCLCPP_WARN(this->get_logger(),
+                        "5Hz Telemetry callback function was called in an "
+                        "invalid state");
+            return;
+        default:
+            break;
+    }
+
+    // Update flight state

```

Comment:
Maybe think about outsourcing each of those messages into a separate file. This makes it easier to change the frequency later on by simply moving the function call from 5 Hz to 10 Hz or even introduce a new frequency if necessary.

Reply:
Fixed in d44c78b
, 4e627f3, 58f11b3, d874c3c